### PR TITLE
fix: [PL-31586]: Migrate ng secrets from local to gcp_kms in prod3

### PR DESCRIPTION
**Pull request by VikasMaddukuriHarness**
## Describe your changes
We are going to introduce gcp_kms as default SM in ng in prod3. So we have migration which should run before this to move all the local Sm to gcp_kms. Once that is done, we will have to move all the secrets that are locally encrypted till now to the new GCP_KMS Sm's. 
This migration handles this movement of secrets from local to gcp_kms. We made this migration behind a FF. 

Testing:
1. Tested the secrets migration locally by creating a gcp_kms and making it as default and then turning this iterator on and verfied the migation of those secrets.
2. Tested the same in pr-env as well. Verified the iterator working based on FF.

TestSteps:
1. created a secret file (gcpKms creds). This is locally encrypted.
2. Created gcpKms sm using this file. Did a testConnection and also created a new secret with this SM to store vault root token. 
3. Now ran the migration. Now the secretfile is turned into gcpKms encryption.
4. Did a testConnection for the Gcp-kms sm. It is success. Created a vault using the created secret (in step2). Validated the vault connection. It is also success.
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-ALL: `trigger tiAll`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44050)

### CODE REVIEW

File: 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@OwnedBy(HarnessTeam.PL)
+public class NgLocalToGlobalSecretsMigrationJob implements Managed {
+  private Future<?> ngLocalToGlobalSecretsMigrationFuture;
+  private final ScheduledExecutorService executorService;
+  private static final String DEBUG_MESSAGE = "DefaultUserGroupsCreationJob: ";
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java:28]: change name to NgLocalToGlobalSecretsMigrationJob

File: 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    String threadName = "ng-secrets-migration-local-to-GcpKms-thread";
+    this.executorService =
+        Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder().setNameFormat(threadName).build());
+  }
+
+  @Override
+  public void start() throws Exception {
+    log.info(DEBUG_MESSAGE + "started...");
+    Random random = new Random();
+    int delay = random.nextInt(2) + 2;
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java:44]: lets keep this as 15/10 mins

File: 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    ngLocalToGlobalSecretsMigrationFuture =
+        executorService.scheduleWithFixedDelay(ngLocalToGlobalSecretsMigrationService, delay, 100, TimeUnit.MINUTES);
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java:46]: also interval to 12 hrs

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java [hunk 4]
```diff
@@ -627,7 +627,9 @@ public enum FeatureName {
       "Do not add quotes to strings when a user reconciles a template, pipeline", HarnessTeam.CDC, Scope.GLOBAL),
   SSCA_ENABLED("FF to enable SSCA on Harness", HarnessTeam.SSCA),
   PL_JPMC_SCIM_REQUIREMENTS(
-      "This flag relates to all the changes requested by JPMC towards SCIM API calls", HarnessTeam.PL);
+      "This flag relates to all the changes requested by JPMC towards SCIM API calls", HarnessTeam.PL),
+  MIGRATE_SECRETS_FROM_LOCAL_TO_GCP_KMS_PROD3(
```
Comment (meenakshi-harness) [inline 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:631]: Rename to PL_MIGRATE_SECRETS_FROM_LOCAL_TO_GCP_KMS_PROD3

File: 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    this.mongoTemplate = mongoTemplate;
+    this.secretsFileService = secretsFileService;
+    this.ngEncryptedDataService = ngEncryptedDataService;
+    this.ngEncryptedDataDao = ngEncryptedDataDao;
+    this.ngFeatureFlagHelperService = ngFeatureFlagHelperService;
+  }
+
+  @Override
+  public void run() {
+    log.info("Started migration of Ng secrets from local to Gcp_kms");
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java:63]: add lock acquire logic

File: 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    CloseableIterator<NGEncryptedData> iterator = runQueryWithBatch(getAllNgEncryptedRecords());
+    while (iterator.hasNext()) {
+      NGEncryptedData ngEncryptedData = iterator.next();
+      log.info("Migrating secret {} from local to gcp", ngEncryptedData.getId());
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java:67]: log accountId with all logs

File: 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+      try {
+        processRecord(ngEncryptedData);
+        log.info("Successfully migrated secret {} from local to gcp", ngEncryptedData.getId());
+      } catch (Exception ex) {
+        log.error("migration failed for encrypted record {} with this exception {}", ngEncryptedData.getId(), ex);
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java:72]: same

**Review by VikasMaddukuriHarness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44050)* status: 6 file(s) to review. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java` line 28 at r5](https://reviewable.io/reviews/harness/harness-core/44050#-NOP-sQfzyi0B_t-YHWD-r5-28:-NPHNI1o5Y9YdKMcfc8i:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e3ded5f1bf8a4975373a34396c154f5fdd25e9da/120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java#L28)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

change msg to NgLocalToGlobalSecretsMigrationJob
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java` line 44 at r5](https://reviewable.io/reviews/harness/harness-core/44050#-NOP-sQfzyi0B_t-YHWD-r5-44:-NPHNIM46rEx6VUuxn98:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e3ded5f1bf8a4975373a34396c154f5fdd25e9da/120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java#L44)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

lets keep this as 15/10 mins
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java` line 46 at r5](https://reviewable.io/reviews/harness/harness-core/44050#-NOP-sQfzyi0B_t-YHWD-r5-46:-NPHNIso3cUeXFrXiXNj:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e3ded5f1bf8a4975373a34396c154f5fdd25e9da/120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationJob.java#L46)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

also interval to 2 hrs
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java` line 63 at r5](https://reviewable.io/reviews/harness/harness-core/44050#-NOP-sQfzyi0B_t-YHWE-r5-63:-NPHNJQ34tKsN_sM8mgN:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e3ded5f1bf8a4975373a34396c154f5fdd25e9da/120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java#L63)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

add lock acquire logic
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java` line 67 at r5](https://reviewable.io/reviews/harness/harness-core/44050#-NOP-sQfzyi0B_t-YHWE-r5-67:-NPHNKKH1vcdm4dMaNmb:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e3ded5f1bf8a4975373a34396c154f5fdd25e9da/120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java#L67)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

log accountId with all logs
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java` line 72 at r5](https://reviewable.io/reviews/harness/harness-core/44050#-NOP-sQfzyi0B_t-YHWE-r5-72:-NPHNKpc0bvnV09Ns5Cv:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e3ded5f1bf8a4975373a34396c154f5fdd25e9da/120-ng-manager/src/main/java/io/harness/ng/core/migration/NgLocalToGlobalSecretsMigrationService.java#L72)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.
___
*[`956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java` line 631 at r5](https://reviewable.io/reviews/harness/harness-core/44050#-NOx0WYsSVXzauChZjtP-r5-631:-NPHNLMX-71jjXSSteBj:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e3ded5f1bf8a4975373a34396c154f5fdd25e9da/956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java#L631)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

Rename to PL_MIGRATE_SECRETS_FROM_LOCAL_TO_GCP_KMS_PROD3
</blockquote></details>

Done.

**Review by shashank-harness — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 1 of 3 files at r6, 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44050)* status: 7 file(s) to review. 7. unresolved discussions. 1 of 2 approvals obtained.

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/utils/RESTToGraphQLHelper.java [hunk 8]
```diff
@@ -465,26 +465,26 @@ private static HashMap<String, QLCEViewFieldInput> getViewFieldsHashMap() {
             .identifierName(ViewFieldIdentifier.CLUSTER.getDisplayName())
             .build());
     viewFieldHashMap.put("CLUSTER_NODE",
-            QLCEViewFieldInput.builder()
-                    .fieldId(clusterFieldsNg.get(6).getFieldId())
-                    .fieldName(clusterFieldsNg.get(6).getFieldId())
-                    .identifier(ViewFieldIdentifier.CLUSTER)
-                    .identifierName(ViewFieldIdentifier.CLUSTER.getDisplayName())
-                    .build());
+        QLCEViewFieldInput.builder()
```
Comment (meenakshi-harness) [inline ce-nextgen/service/src/main/java/io/harness/ccm/remote/utils/RESTToGraphQLHelper.java:468]: why this formatting?

**Pull request closed without merging**